### PR TITLE
[OM-89077]: New cpu throttling metrics computation and configurable VCPU Throttling Threshold

### PR DESCRIPTION
**Pull request by pallavidn**
Addresses these two server side jira tasks.

**[OM-91748](https://vmturbo.atlassian.net/browse/OM-91748): calculate the new throttling metrics**
For the reasons listed in the parent story the new calculation of the throttling metrics is as follows:
For each sampling window:
`container_cpu_cfs_throttled_seconds_total increase * 100/container_cpu_usage_seconds_total increase`

In this review added changes as follows:
-  kubelet to collect 2 new metrics 
`container_cpu_cfs_throttled_seconds_total: Total time duration the container has been throttled`
`container_cpu_usage_seconds_total: Cumulative cpu time consumed in seconds`

- Modify the `aggregateContainerThrottlingSamples` in `general_builder.go` to compute the throttling time, total cpu usage time over the samples collected in between each discovery interval
- Modified the `container_spec_entity_dto_builder.go` to compute the average throttled percent to use the new metrics.

**[OM-91752](https://vmturbo.atlassian.net/browse/OM-91752) configurable util threshold (default to 30%)**
VCPU Throttling threshold was hard-coded to 5%
With this new measurement, there’s no longer a need to have an extremely low throttling desired state. 
- Setting the new default the util threshold to be 30%
- In addition this setting is configurable

### Testing Done:
- Ran kubeturbo with the new throttling computations pointing to the PT-dc-11 cluster which contains workloads deployed specifically testing for cpu throttling related resizes.
Compared with another appliance with a current kubeturbo pointing to the same cluster.
- Collected raw metrics and analyzed the new and old throttling averages per discovery period 
[cpu-throttling.xlsx](https://github.com/turbonomic/kubeturbo/files/10006200/cpu-throttling.xlsx)

### CODE REVIEW

File: pkg/discovery/dtofactory/general_builder.go [hunk 1]
```diff
@@ -401,35 +405,28 @@ func aggregateContainerThrottlingSamples(entityID string, samples []metrics.Thro
...
-				throttledPercent = throttledSingleSample * 100 / totalSingleSample
-			}
+		throttledTimePercent := float64(0)
+		throttledTimeSingleSample := samples[i+1].ThrottledTime - samples[i].ThrottledTime
+		totalUsageSingleSample := samples[i+1].TotalUsage - samples[i].TotalUsage
+		if throttledTimeSingleSample > 0 || totalUsageSingleSample > 0 {
+			throttledTimePercent = throttledTimeSingleSample * 100 / (throttledTimeSingleSample + totalUsageSingleSample)
 		}
-		peak = math.Max(peak, throttledPercent)
+		peakTimeThrottled = math.Max(peakTimeThrottled, throttledTimePercent) //new
```
Comment (chlam4) [inline pkg/discovery/dtofactory/general_builder.go:423]: Nit: should remove the unnecessary trailing comment?

File: pkg/discovery/dtofactory/general_builder.go [hunk 2]
```diff
@@ -401,35 +405,28 @@ func aggregateContainerThrottlingSamples(entityID string, samples []metrics.Thro
 
 	lastReset := 0
 	for i := 0; i < numberOfSamples-1; i++ {
-		if samples[i+1].Total < samples[i].Total || samples[i+1].Throttled < samples[i].Throttled {
+		if samples[i+1].TotalUsage < samples[i].TotalUsage || samples[i+1].ThrottledTime < samples[i].ThrottledTime {
 			// This probably means the counter was reset for some reason
-			throttled += (samples[i].Throttled - samples[lastReset].Throttled)
-			total += samples[i].Total - samples[lastReset].Total
+			throttledTime += (samples[i].ThrottledTime - samples[lastReset].ThrottledTime)
```
Comment (chlam4) [inline pkg/discovery/dtofactory/general_builder.go:410]: Nit: the parantheses are not necessary.

File: pkg/discovery/dtofactory/general_builder_test.go [hunk 3]
```diff
@@ -332,17 +344,20 @@ func TestMetricValueWithThrottlingCumulativePoints(t *testing.T) {
...
+	fmt.Printf("metricValue %v\n", metricValue)
 	// time                  t1   t2    t3     t4      t5                  t6     t7    t8               t9     t10    t11
 	// throttled samples =   1    3     5      6       11                  0      1     3                1       3      1
 	// total samples =       5    8     10     15      25                  0      5     8                5       8      5
 	// diff                   d1    d2     d3      d4          d5              d6    d7       d8              d9       d10
 	// throttled =           (3-1) (5-3)  (6-5)   (11-6)  (reset/ignore 0-11) (1-0) (3-1) (reset/ignore 1-3) (3-1)  (reset/ignore 1-3)
 	// total =               (8-5) (10-8) (15-10) (25-15) (reset/ignore 0-25) (5-0) (8-5) (reset/ignore 5-8) (8-5)  (reset/ignore 5-8)
+	// avg =				 2/2+3  2/2+2   1/1+5   5/5+10						1/1+5 2/2+3						2/2+3
+	//						  time = (11-1) usage = (25-5)                  time = (3-0) usage =(8-0)		time=(3-1) usage=(8-5)
+	//						  40	50	    16.67   33.33                       16.67  40                        40
```
Comment (chlam4) [inline pkg/discovery/dtofactory/general_builder_test.go:356]: These 3 new rows are not aligned with the previous ones in terms of spacing.  Is it what you intend to have?

**Review by chlam4 — APPROVED** [review]
Nice work @pallavidn!  I'm approving it with just a few nit comments for you to consider.

File: pkg/discovery/dtofactory/container_spec_entity_dto_builder.go [hunk 4]
```diff
@@ -167,31 +171,27 @@ func (builder *containerSpecDTOBuilder) getCommoditiesSold(containerSpecMetrics
...
-			avgThrottledOverall = throttledOverall * 100 / totalOverall
-		}
+
+	avgThrottledTimeOverall := float64(0)
+	if throttledTimeOverall > 0 || totalUsageOverall > 0 {
+		avgThrottledTimeOverall = throttledTimeOverall * 100 / (throttledTimeOverall + totalUsageOverall)
 	}
 
-	return 100, avgThrottledOverall, peakOverall
+	return 100, avgThrottledTimeOverall, peakThrottledPercentOverall //avgThrottledOverall, peakOverall
```
[the remark this answers is not in the document]
Comment (pallavidn) [reply pkg/discovery/dtofactory/container_spec_entity_dto_builder.go]: The other variable is the peak throttled percent.

**Pull request merged**